### PR TITLE
Remove uag next

### PR DESCRIPTION
**Pull request by cspiel1**
Get rid of the current UA which was used for dialing and also for answer call, hangup.

Instead menu holds the id of the active call.

TODO: For dialing an automatic UA selection will be implemented.

### CODE REVIEW

**Comment by cspiel1** [comment]
@sreimers This fails:
```
sudo apt-get update && sudo apt-get install -y libssl-dev valgrind
```
Should I solve this somehow?

**Comment by sreimers** [comment]
> @sreimers This fails:
> ```
> sudo apt-get update && sudo apt-get install -y libssl-dev valgrind
> ```
> Should I solve this somehow?

I just re-run and looks good now. Rarely the used apt mirrors seems to be broken.

**Comment by cspiel1** [comment]
TODO: The call state has to be taken into account somehow for the "active call". Relates also to @alfredh-s comment https://github.com/baresip/baresip/pull/1199#issuecomment-769947402

**Comment by alfredh** [comment]
the ua context can be passed in from the command input, for example mqtt:

https://github.com/baresip/baresip/blob/master/modules/mqtt/subscribe.c#L83

it is therefore important that the ua selection logic be kept.

File: modules/menu/dynamic_menu.c [hunk 1]
```diff
@@ -12,76 +12,49 @@
 static int set_audio_bitrate(struct re_printf *pf, void *arg)
 {
 	struct cmd_arg *carg = arg;
-	struct ua *ua = carg->data ? carg->data : menu_uacur();
-	struct call *call;
+	struct call *call = menu_getcall(pf);
```
Comment (alfredh) [inline modules/menu/dynamic_menu.c:15]: we should keep this logic:

1. if carg->data is set, use ua from there
2. if carg->data is NULL, lookup ua from current call

File: modules/menu/menu.c [hunk 2]
```diff
@@ -580,18 +582,77 @@ struct menu *menu_get(void)
...
+ */
+int menu_selcall(struct re_printf *pf, const char *id)
 {
-	if (!menu.ua_cur)
-		menu.ua_cur = list_ledata(list_head(uag_list()));
+	struct call *call = menu_find_call(id);
+	struct list *calls = ua_calls(call_get_ua(call));
+
+	if (call) {
+		menu.callid = id;
```
Comment (alfredh) [inline modules/menu/menu.c:617]: please note that the call-id string is dynamically allocated, and should be
copied using e.g. `str_dup`.

otherwise if the call object is destructed, the callid string will point to freed memory

File: modules/menu/menu.h [hunk 3]
```diff
@@ -17,8 +17,7 @@ struct menu{
 	struct tmr tmr_stat;          /**< Call status timer              */
 	struct play *play;            /**< Current audio player state     */
 	struct mbuf *dialbuf;         /**< Buffer for dialled number      */
-	struct le *le_cur;            /**< Current User-Agent (struct ua) */
-	struct ua *ua_cur;
+	const char *callid;           /**< Call-id of active call         */
```
Comment (alfredh) [inline modules/menu/menu.h:20]: should be `char *callid`

also remember to mem_deref in module_close().

**Comment by alfredh** [comment]
I can suggest some simple code for setting/getting active call:

```c
void menu_selcall(struct call *call)
{
	menu.callid = mem_deref(menu.callid);

	if (call)
		str_dup(&menu.callid, call_id(call));
}


struct call *menu_getcall(void)
{
	if (menu.callid)
		return menu_find_call(menu.callid);
	else
		return NULL;
}


struct ua *menu_uacur(void)
{
	return call_get_ua(menu_getcall());
}
```

for `CALL_CLOSED` event you can call menu_setcall(NULL);

if you keep the menu_uacur() signature, it will also make the diff smaller :)

File: modules/menu/dynamic_menu.c (hunk 1 above)
Comment (cspiel1) [reply modules/menu/dynamic_menu.c:15]: Oh, I see. Did'nt saw how cmd_process_long() is used in mqtt. In ctrl_tcp and ctrl_dbus we should do this in the same way. 

But here is a problem. The parameter is a void*. And you allow all commands. Suppose one writes a command "do_foo" and wants to get a foo* in the data parameter and maybe he writes also an ui module which correctly passes a foo* in the data parameter. Then a user may also call do_foo by means of mqtt and specify the accountaor. Then ua* is cast to foo*.

**Comment by cspiel1** [comment]
In menu_selcall() you omitted
```
call_set_current(calls, call);
```
We don't need the reorder of the calls in the UA, right?

File: modules/menu/dynamic_menu.c (hunk 1 above)
Comment (cspiel1) [reply modules/menu/dynamic_menu.c:15]: For now I'll keep the logic.

**Comment by cspiel1** [comment]
Because you want to keep the ua selection logic per data parameter (https://github.com/baresip/baresip/pull/1207#discussion_r573220048) we also have to call call_set_current(). In order to let the diff be small like you stated, I will replace the last commit.

**Comment by cspiel1** [comment]
The last commit is a little bit off topic. But came up to me while integrating the new menu_find_call(). Should I make a separate PR for this?

**Comment by cspiel1** [comment]
Also what fits to this PR is what @alfredh already suggested: cmd_answer() should operate on the last incoming call.

- first hold the last established call
- then answer the last incoming call

**Comment by alfredh** [comment]
I tested calling a number and it is not working:

```
/dial music
menu: ua_connect failed: Invalid argument
```

File: modules/menu/dynamic_menu.c [hunk 4]
```diff
@@ -15,19 +15,19 @@ static int set_audio_bitrate(struct re_printf *pf, void *arg)
...
 		audio_set_bitrate(call_audio(call), bitrate);
 	}
 	else {
-		err = re_hprintf(pf, "call not found\n");
+		(void)re_hprintf(pf, "call not found\n");
+		return EINVAL;
 	}
 
-	return err;
+	return 0;
```
Comment (alfredh) [inline modules/menu/dynamic_menu.c:30]: can you please make a separate PR for this block ?

it is not related to the main topic of this PR.

File: modules/menu/static_menu.c [hunk 5]
```diff
@@ -116,7 +115,7 @@ static int cmd_answerdir(struct re_printf *pf, void *arg)
 			"[^ ]*", &argdir[0]);
 
 	if (err) {
-		warning("%s", usage);
+		(void) re_hprintf(pf, "%s", usage);
```
Comment (alfredh) [inline modules/menu/static_menu.c:118]: please also make a separate PR for this and similar changes.

**Comment by alfredh** [comment]
can you please rebase this PR on top of master HEAD ?

also, is it possible for you to squash all the commits into one single commit,
since you are doing a forced push ? it will make things a bit cleaner.

**Comment by juha-h** [comment]
Sorry, but I haven't really followed the details of "Remove uag next"
effort.

What is the replacement of it?

Let's say that I have two accounts sip:foo@example.com and
sip:bar@another.com.  Then I want to make a call to PSTN:

/dial +3582345670

How can I choose, which of the two accounts is selected to make the
call, since there is nothing that could make baresip to automatically
choose the right one?

**Comment by alfredh** [comment]
I would like to suggest that the UA selection is done on a per-command basis.

example:

```
/dial +3582345670 ua=0
```

**Comment by juha-h** [comment]
Alfred E. Heggestad writes:
> I would like to suggest that the UA selection is done on a per-command basis.
> 
> example:
> 
> ```
> /dial +3582345670 ua=0
> ```

Then I would need to remember, which one is ua 0.

And how about if I chat with someone?  Would I need in every message
also tell the ua?

**Comment by cspiel1** [comment]
- PR https://github.com/baresip/baresip/pull/1217 should do the automatic selection. An PSTN would be handled like a user only SIP address. This would select the first registered UA.
- A manual selection like @alfredh suggests as second command parameter would replace the uag_next() if the application wants to select a specific UA. --> TODO: another PR.

A chat application may invoke:
- /message a@proxy1
- /message a ua=0

A pre-selection could be handled in the application.

**Comment by cspiel1** [comment]
> I tested calling a number and it is not working:
> 
> ```
> /dial music
> menu: ua_connect failed: Invalid argument
> ```

This PR removes uag_next() and introduces the active call for commands that belong to the active call.
The UA selection for dial is provided by https://github.com/baresip/baresip/pull/1217.

**Comment by juha-h** [comment]
Currently I can re-dial previous destination just by pressing `d`.  If it is menu module that remembers the previous destination, could it do the same for ua?

**Comment by juha-h** [comment]
Just checked and there exists le_cur, ua_cur, and redial_aor fields in `struct menu`.  So if I select one of the uas as current in menu module, there is no need for current ua related API calls and user experience would not need to change.  It should thus be possible to keep the /ua_next command (if there ever was intention to remove it).

**Comment by cspiel1** [comment]
The ua_cur was the UA of the active call. And this UA also was used for future dials. But I think this are two different use-cases.

In the closed PR https://github.com/baresip/baresip/pull/1199#discussion_r566977088 @alfredh suggested to remove the current UA. The plan is to:
- Remove the current UA for dialing. Instead remember the active call for commands like answer, hold, resume, ...
- Add an automatic UA selection for dial, options and any outgoing requests.
- Add an optional manual UA selection for outgoing commands that overrides the automatic UA selection. It should be more convenient then ua_next().
- The redial_aor is used for scode 701. But we could also use it for a manual redial and link the 'd' in menu to a new function.

This PR is only the first part. The second is https://github.com/baresip/baresip/pull/1217. And I see that it was not a good idea to make separate PRs. I will add the commits of https://github.com/baresip/baresip/pull/1217 here. Maybe it will make clearer that we want to improve the comfort for the user, instead of removing functionality. Then I will also add a manual override.

But I suggest to review the commits one-by-one. This should be easier.

**Comment by cspiel1** [comment]
@alfredh @juha-h Now this PR includes also the automatic UA selection and the manual override for dial and options. The indices for the manual selection are printed with command reginfo.

The manual override (function menu_ua_carg) is a first draft. Your suggestions are welcome.

For future PRs we may use the manual override also for other commands.

@juha-h You mentioned the UA selection for cmd_message. This is part of module contact. Nothing changes for this command, because it uses the unchanged ua_lookup_domain() in module contact.

**Comment by juha-h** [comment]
Christian Spielberger writes:

> @juha-h You mentioned the UA selection for cmd_message. This is part
> of module contact. Nothing changes for this command, because it uses
> the unchanged ua_lookup_domain() in module contact.

Good, since I consider >/< (contact_next/contact_prev) very useful.
After all the enhancement, will there exist similar ua_next/ua_prev menu
commands?

**Comment by cspiel1** [comment]
uag_next() is not used for cmd_message. Instead ua_lookup_domain().

In an intermediate state of PR https://github.com/baresip/baresip/pull/1199 I proposed an "auto" mode which could be selected with uag_next() by iterating over the UAs. If "auto" was selected then the automatic UA selection was used. But only for dial. @alfredh and me want to get rid of uag_next(). I think that the user should not have to deal with selecting an adequate UA. He wants to specify a SIP target address, not more. Then it should just work.

The advanced user may use the manual per command UA selection, which I added here shortly as a first draft.

First of all, we should distinguish the active call UA from the outgoing request UA (like dial). @alfredh Please could you make a decision?
- Should we keep uag_next() for outgoing requests? This would be ok for me, if we do this in addition to the automatic selection.
- Should we also keep the proposed manual per command UA selection, like you suggest above?
